llama_index/callbacks/promptlayer_handler.py

```diff
@@ -0,0 +1,117 @@
+                "Please install PromptLAyer with `pip install promptlayer`"
+            )
+        self.pl_tags = pl_tags
+        self.return_pl_id = return_pl_id
+        super().__init__(event_starts_to_ignore=[], event_ends_to_ignore=[])
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        return
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        return
+
+    event_map: Dict[str, Dict[str, Any]] = {}
+
+    def add_event(self, event_id: str, **kwargs: Any) -> None:
+        self.event_map[event_id] = {
+            "kwargs": kwargs,
+            "request_start_time": datetime.datetime.now().timestamp(),
+        }
+
+    def get_event(
+        self,
+        event_id: str,
+    ) -> Dict[str, Any]:
+        return self.event_map[event_id] or {}
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        parent_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        if event_type == CBEventType.LLM and payload is not None:
+            self.add_event(
+                event_id=event_id, **payload.get(EventPayload.SERIALIZED, {})
+            )
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        if event_type != CBEventType.LLM or payload is None:
+            return
+        request_end_time = datetime.datetime.now().timestamp()
+        prompt = str(payload.get(EventPayload.PROMPT))
+        completion = payload.get(EventPayload.COMPLETION)
+        response = payload.get(EventPayload.RESPONSE)
+        function_name = PROMPT_LAYER_CHAT_FUNCTION_NAME
+        event_data = self.get_event(event_id=event_id)
+        resp: str | Dict
+        extra_args = {}
+        if response:
+            messages = cast(List[ChatMessage], payload.get(EventPayload.MESSAGES, []))
+            resp = response.message.dict()
+            extra_args = {
+                "messages": [message.dict() for message in messages],
+            }
+            ## promptlayer needs tool_calls toplevel.
+            if "tool_calls" in response.message.additional_kwargs:
+                resp["tool_calls"] = [
+                    tool_call.dict()
+                    for tool_call in resp["additional_kwargs"]["tool_calls"]
+                ]
+                del resp["additional_kwargs"]["tool_calls"]
+        if completion:
+            function_name = PROMPT_LAYER_COMPLETION_FUNCTION_NAME
+            resp = str(completion)
+        pl_request_id = self._promptlayer_api_request(
+            function_name,
+            "openai",
+            [prompt],
+            {
+                **extra_args,
+                **event_data["kwargs"],
```

Comment:
I think prompt layer can support token usage, got to figure out how to pass that in as well, does openai respond w/ token usage or do we need to use token_counter callback?

Reply:
You dont not need to send token count, if you have it readily available you send send them the same way openai sends them which is 
```
"usage": {
    "prompt_tokens": 9,
    "completion_tokens": 12,
    "total_tokens": 21
  }
```
in the function kwargs